snuba/stateful_consumer/states/bootstrap.py

```diff
@@ -14,12 +22,60 @@ class BootstrapState(State[ConsumerStateCompletionEvent, ConsumerStateData]):
     control topic.
     """
 
+    def __init__(self,
+        topic: str,
+        bootstrap_servers: Sequence[str],
+        group_id: str,
+    ):
+        super(BootstrapState, self).__init__()
+
+        def on_partitions_assigned(
+            consumer: Consumer,
+            partitions: Sequence[TopicPartition],
+        ):
+            pass
+
+        def on_partitions_revoked(
+            consumer: Consumer,
+            partitions: Sequence[TopicPartition],
+        ):
+            pass
+
+        self.__consumer = StrictConsumer(
+            topic=topic,
+            bootstrap_servers=bootstrap_servers,
+            group_id=group_id,
+            auto_offset_reset="earliest",
```

Comment:
In this case, shouldn't automatic offset resets be disallowed, since they could lead to some potentially dangerous behavior? (I might be thinking too catastrophically here, let me know if this is actually safe.)

If this happens to either commit an offset that is too far in the future (not sure if this possible due to replication semantics during a failover, and very unlikely to happen by manually committing the wrong offsets, but possible) or the current offset falls off retention, this would lead to the offset being reset silently. I think we'd want to at least know that's the case if it does happen — even if the response is to just manually move the offset to the earliest point?

The SynchronizedConsumer deals with this by setting the `auto.offset.reset`, and dealing with setting it's own start point in the assignment callback, but only if the consumer group has never been previously registered for that topic (or has fallen out of retention in the Kafka consumer offsets topic): https://github.com/getsentry/sentry/blob/c9cca5dc1120f0e30838b4f1c1d5a0dce32c174b/src/sentry/eventstream/kafka/consumer.py#L270

Reply:
_In this case, shouldn't automatic offset resets be disallowed_
How do I do that?

Yes, resetting offset should not be allowed. If offsets that were not yet consumed drop off retention this should at least require a human action. At that point we would just restart from earliest hoping there is a new snapshot above still within retention. But more in general (beyond snapshots) I agree the user should be notified if it cannot read the whole topic from the previous committed offset

Reply:
> How do I do that?

Set `auto.offset.reset` to `error`, and then manually set the offsets based on the watermark offsets in the assignment callback. The SynchronizedConsumer example above should be relatively straightforward, but could probably be commented better.

Reply:
Actually I am not sure this would be the right behavior.
The goal of this consumer is: pick up a topic (single partition). Consume it from wherever we can to the end once. We can fail this process in case of partition reassignment while this is happening and restart safely.
Since this is supposed to consume something only during snapshots, consuming will most of the time raise OFFSET_OUT_OF_RANGE. Which would require us to consume from earliest.